### ISSUE TITLE
Default values for init-only properties are lost with generated JsonTypeInfo

### ISSUE THREAD

**Issue by Sergio0694** [dotnet/runtime#84484]
### Description

> **Note**: related to #58770 

With the 8.0.0 release, System.Text.Json will support required and init-only properties. This is something we'd like to leverage in the Microsoft Store, as moving away from records to instead use normal classes with required init-only properties can help reduce code size (due to no longer having all the record fluff), while still keeping things "well defined" (ie. not having nullable properties you need to suppress warnings for, etc.). But, I've noticed that default values don't really seem to work.

Consider this example:

```csharp
Foo? foo = JsonSerializer.Deserialize("{}", MyJsonContext.Default.Foo);

Console.WriteLine(foo);

[JsonSerializable(typeof(Foo))]
public partial class MyJsonContext : JsonSerializerContext
{
}

public record Foo(string Name = "DefaultName", int Number = 42);
```

This will print "Foo { Name = DefaultName, Number = 42 }". That is, those default values are correctly used.

Consider the same thing, but changing that record to a class like this:

```csharp
public class Foo
{
    public string Name { get; init; } = "DefaultName";

    public int Number { get; init; } = 42;
}
```

This will cause those default values to be completely lost instead. The same code above will deserialize an instance that just has the values of each property set to `default`. This is not really intuitive - I would've expected the default value from those property initializers to be preserved, whereas the generated code is just always assigning _something_ to every single property, even if the value isn't present in the JSON, which will just cause `default` to be assigned instead.

Shouldn't the generated code for the object initializer only assign init-only, non-required properties when they're actually present in the JSON? If not, is there some way to preserve the "default value" behavior for properties in some way?

cc. @eiriktsarpalis @layomia

### Reproduction Steps

n/a

### Expected behavior

n/a

### Actual behavior

n/a

**Comment by eiriktsarpalis** [comment]
Yeah, this is a known drawback of how the source generator is populating init-only properties. It's a chicken and egg situation: you cannot get the value of an init-only property before the object has been created, and you cannot set the value of an init-only property after the object has been created.

> Shouldn't the generated code for the object initializer only assign init-only, non-required properties when they're actually present in the JSON?

The generator can only create the construction delegate at compile time, whereas what properties are specified in the JSON can only be known at run time. One possible solution would be to generate one compiler delegate for each possible subset of init-only properties are being specified, however that would result in a lot of code being generated: given $n$ init-only properties on a type, you would need to generate $2^n$ constructor delegates.

One obvious solution would be to use reflection, which would be trim/AOT safe in this instance. Another possibility might be that the compiler provides proxy APIs that serializers can use to write to init-only properties.

cc @jaredpar @eerhardt @jkotas @layomia

**Comment by Sergio0694** [comment]
> "Yeah, this is a known drawback of how the source generator is populating init-only properties. It's a chicken and egg situation: you cannot get the value of an init-only property before the object has been created, and you cannot set the value of an init-only property after the object has been created."

Ah... Right. Yeah that's not great 🥲

What about having some `[JsonDefaultValue]` attribute one could use for properties? And the generator could read the value from that and assign the corresponding value in the generated type info, so that it'd match exactly the same behavior you get today when using default values for parameters of the primary record constructor? 🤔

Something like, in the example above:

```csharp
public class Foo
{
    [JsonDefaultValue("DefaultName")]
    public string Name { get; init; }

    [JsonDefaultValue(42)]
    public int Number { get; init; }
}
```

And you'd get the same as with records. The default value would be constrained to constants, but then again that's the same limitation you get for default values in parameters anyway, so it would essentially be on par with that?

**Comment by jkotas** [comment]
> Another possibility might be that the compiler provides proxy APIs that serializers can use to write to init-only properties.

https://github.com/dotnet/runtime/issues/81741

**Comment by Sergio0694** [comment]
I know it's lower priority but I kinda hope the solution for this won't involve that, so it can also work downlevel 🥲
That's why I was hoping we could maybe just use something like an attribute to instruct the generator instead.

**Comment by eiriktsarpalis** [comment]
> > Another possibility might be that the compiler provides proxy APIs that serializers can use to write to init-only properties.
> 
> #81741

Assuming this requires new API, we wouldn't be able to use it in `netstandard2.0` SG targets though. I kind of like the pattern matching on reflection invoke approach, it would work in old TFMs but also be fast in bleeding edge ones.

**Comment by jkotas** [comment]
> we wouldn't be able to use it in netstandard2.0 SG targets though. I kind of like the pattern matching on reflection invoke approach, it would work in old TFMs but also be fast in bleeding edge ones.

We have discussed the pros and cons of each approach in depth. The problem with pattern matching reflection invoke is that it is fragile and very complicated.

netstandard2.0 targets can use classic reflection invoke. There is a lot of performance left on table when targeting netstandard2.0 in general.

**Comment by eiriktsarpalis** [comment]
There's also added complexity for SG's needing to implement different access patterns depending on the TFM. Using inlined reflection everywhere seems like an obvious win for authors.

> The problem with pattern matching reflection invoke is that it is fragile and very complicated.

Is it related to overload resolution?

**Comment by jkotas** [comment]
Here is a simple example of a method invocation that takes integer and string: `typeof(MyType).GetMethod("MyMethod", new Type[] { typeof(int), typeof(string) }).Invoke(null, new object[] { intValue, stringValue });`. This gets compiled into a lot of IL. There are multiple method calls, multiple array allocations, array element assignments, boxing, ... . It is what makes the pattern match complicated and fragile.

Also, if the downlevel SG cared about performance even a bit, they would not want to use this unified pattern. They would want to cache the MethodInfo into a static or even concert it to a delegate, to reduce the steady perf hit.

**Comment by eiriktsarpalis** [comment]
> This gets compiled into a lot of IL.

Assuming the scope of the proposed feature is serializer-like components, then we only really care about field, property and constructor member access. Fields and properties have relatively simple access patterns and supporting constructors isn't essential (since a source generator typically doesn't need to bypass accessibility there).

> They would want to cache the MethodInfo into a static or even convert it to a delegate, to reduce the steady perf hit.

Perhaps we should reconsider `infoof` :-) or something along these lines that avoids visibility checks, implements caching and uses an IL pattern that the JIT can easily recognise.

**Comment by jkotas** [comment]
> supporting constructors isn't essential (since a source generator typically doesn't need to bypass accessibility there).

Constructors with parameters are in the list of EF Core requirements for this feature.

**Comment by Sergio0694** [comment]
Will the generator support primary constructors in C# 12? Because I'm thinking a solution could be to use them in cases where you do need some default values, to still get some lightweight "records" without all the additional metadata bloat. Like:

```csharp
public class Foo(
    [property: JsonRequired] string firstName,
    [property: JsonRequired] string lastName,
    string tag = "DefaultValue",
    int number = 42)
{
    public string FirstName => firstName;
    public string LastName => lastName;
    public string Tag => tag;
    public int Number => number;
}
```

It's not ideal because you'd have to duplicate each property (as in, declare both a primary constructor parameter as well as an actual property), but at least it'd avoid all the additional codegen that normal records would cause. Since System.Text.Json already handles this for record constructors and explicit constructors, I assume this could just work? 🤔

**Comment by eiriktsarpalis** [comment]
> Since System.Text.Json already handles this for record constructors and explicit constructors, I assume this could just work?

I believe so. Neither the source generator nor the reflection serialization have any special provision for positional record, but they still work. The only issue I'm aware of is inability to annotate the primary constructor with attributes such as `JsonConstructorAttribute`, but that is something to be addressed by the language. We should add test coverage for the new feature regardless.

**Comment by Sergio0694** [comment]
I think that might have already been fixed by https://github.com/dotnet/roslyn/pull/67695 😄

**Comment by eiriktsarpalis** [comment]
Moving to Future for now.

**Comment by eiriktsarpalis** [comment]
I expect this will eventually be addressed by replacing the current constructor delegate approach with setters generated via https://github.com/dotnet/runtime/issues/81741

**Comment by Neme12** [comment]
> even convert it to a delegate

I agree, caching the delegate seems like the best option.

**Comment by Neme12** [comment]
> The generator can only create the construction delegate at compile time, whereas what properties are specified in the JSON can only be known at run time. One possible solution would be to generate one constructor delegate for each possible subset of init-only properties are being specified, however that would result in a lot of code being generated: given n init-only properties on a type, you would need to generate 2n constructor delegates.

I understand the issue. But why does it then work for records though? Records also have init-only properties by default and default values still work there, so I'm confused.

**Comment by VAllens** [comment]
Any progress?
We believe the two behaviors should be consistent !
